### PR TITLE
Cleanup dense poly

### PR DESCRIPTION
**Pull request by mmagician**
Slowly preparing for [generalizing](https://github.com/a16z/Lasso/issues/8) the [PCS](https://github.com/a16z/Lasso/pull/66), which would require switching to `MultilinearExtension` from `ark-poly` and deprecating the custom `DensePolynomial` used now.

This PR should be useful regardless of whether the above happens or not:
1. Removes `merge_dual` in favor of calling `merge` on the concatenated slices
2. Removes unused `extend`
3. msm-based functionality (`flag_msm` & `sm_msm`) are not methods on `DensePolynomial` but rather stand-alone functions. Moves them to the `msm` module.

### CODE REVIEW

**Comment by mmagician** [comment]
Dev on `ark-poly` in that direction: https://github.com/arkworks-rs/algebra/pull/763

File: jolt-core/src/jolt/vm/instruction_lookups.rs [hunk 1]
```diff
@@ -122,7 +122,7 @@ where
     fn batch(&self) -> Self::BatchedPolynomials {
         use rayon::prelude::*;
         let (batched_dim_read, (batched_final, batched_E, batched_flag)) = rayon::join(
-            || DensePolynomial::merge_dual(self.dim.as_ref(), self.read_cts.as_ref()),
+            || DensePolynomial::merge(&[self.dim.as_slice(), self.read_cts.as_slice()].concat()),
```
Comment (sragss) [inline jolt-core/src/jolt/vm/instruction_lookups.rs:125]: Can we drop the concat here?

I'm concerned that we're going to allocate a vector of length `self.dim.len() + self.read_cts.len()` twice. Once after the `.concat()` call and once in `DensePolynomial::merge`. Recall that these are likely to be massive (2^28 -> 2^32).

**Review by sragss — COMMENTED** [review]
One comment on allocation. Otherwise LGTM.

**Comment by sragss** [comment]
LGTM

**Pull request merged**